lib/collection/src/shards/shard_holder/resharding.rs

```diff
@@ -364,6 +369,107 @@ impl ShardHolder {
+        // │ Non-Target │◄──┤ Target   │
+        // │ Receiver   │   │ Sender   │
+        // │            │   │          │
+        // └────────────┘   └──────────┘
+
+        // Target shard of the resharding operation. This is the shard that:
+        //
+        // - *created* during resharding *up*
+        // - *deleted* during resharding *down*
+        let is_target_shard = shard_id == state.shard_id;
+
+        // Shard that will be *receiving* migrated points during resharding:
+        //
+        // - *target* shard during resharding *up*
+        // - *non* target shards during resharding *down*
+        let is_receiver_shard = match state.direction {
+            ReshardingDirection::Up => is_target_shard,
+            ReshardingDirection::Down => !is_target_shard,
+        };
+
+        // Shard that will be *sending* migrated points during resharding:
+        //
+        // - *non* target shards during resharding *up*
+        // - *target* shard during resharding *down*
+        let is_sender_shard = !is_receiver_shard;
+
+        // We split update operations:
+        //
+        // - on *receiver* shards during `MigratingPoints` stage (for all operations except `upsert`)
+        // - and on *sender* shards during `ReadHashRingCommitted` stage when resharding *up*
+
+        let should_split_receiver = is_receiver_shard
+            && state.stage == ReshardStage::MigratingPoints
+            && !operation.is_upsert_points();
+
+        let should_split_sender = is_sender_shard
+            && state.stage >= ReshardStage::ReadHashRingCommitted
+            && state.direction == ReshardingDirection::Up;
+
+        if !should_split_receiver && !should_split_sender {
+            return operation.into();
+        }
+
+        // There's no point splitting delete operations
+        if operation.is_delete_points() {
+            return operation.into();
+        }
+
+        let filter = self
+            .resharding_filter()
+            .expect("resharding filter is available when resharding is in progress");
+
+        let point_ids = operation.point_ids();
+
+        if point_ids.is_empty() {
+            return operation.into();
+        }
+
+        let target_point_ids: HashSet<_> = point_ids
+            .iter()
+            .copied()
+            .filter(|&point_id| filter.check(point_id))
+            .collect();
+
+        if target_point_ids.is_empty() {
+            operation.into()
+        } else if target_point_ids.len() == point_ids.len() {
+            OperationsByMode::default().update_existing(operation)
+        } else {
+            let mut update_all = operation.clone();
+            update_all.retain_point_ids(|point_id| !target_point_ids.contains(point_id));
+
+            let mut update_existing = operation;
+            update_existing.retain_point_ids(|point_id| target_point_ids.contains(point_id));
+
+            OperationsByMode::from(update_all).update_existing(update_existing)
+        }
```

Comment:
We need to extract all "target" point IDs from the operation into a separate "update existing" operation.

Conceptually, this is a lot like `split_by_shard`, but it was too annoying to copy-and-tweak `split_by_shard` as `split_by_mode` or to try to generalize `split_by_shard`/`split_by_mode` into single method.

So I've made a less optimal, but simpler impl based on `get_point_ids` and `retain_point_ids` methods: if operation have to be split in two, we just clone it and then do `retain_point_ids` on both copies.